### PR TITLE
add cpu feature detection for Arm64

### PR DESCRIPTION
**Pull request by lubinsz**
Signed-off-by: Bin Lu <bin.lu@arm.com>

### CODE REVIEW

**Comment by lubinsz** [comment]
@klauspost 
Hi @klauspost ,

Why Travis report this issue?
Log:
-----------------------------------------------------------
The command "go test -race ./..." exited with 0.
0.13s$ diff <(gofmt -d .) <("")
: command not found
------------------------------------------------------------
It seems that no command was found in Travis. Is there any problem in your travis test script?

**Comment by klauspost** [comment]
The problem is that your code doesn't appear to be go fmt'ed. Is that correct?

**Comment by lubinsz** [comment]
Hi @klauspost 
Please take a look at this pr. 

With this patch, I enabled feature detection for Arm64 platform. 
Thanks.

File: detect_arm64.go [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+/*
+ * A little tricky here.
+ * There is no such registers of cpuid in Arm platform.
+ * So we only borrow the name of these functions.
+ *
+ * In armCpuid(), eax = hwcap
+ */
+func armCpuid(op uint32) (eax, ebx, ecx, edx uint32) {
+	eax = uint32(cpuInfo_arm64.HwCap)
+	ebx = 0
```
Comment (klauspost) [inline detect_arm64.go]: These are always 0, no need to set them

File: detect_arm64.go [hunk 2]
```diff
@@ -0,0 +1,246 @@
+// Copyright (c) 2018 Robin Lu (lubinsz@gmail.com), released under MIT License. See LICENSE file.
+// +build arm64,linux
```
Comment (klauspost) [inline detect_arm64.go]: AFAICT `cgo` should also be added as a build constraint.

File: detect_arm64.go [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+	Revision uint64
+	Serial   []byte
+	HwCap    uint64
+}
+
+var (
+	cpuInfo_arm64 *cpuinfo_arm
+)
+
+func CPUInit_arm() {
```
Comment (klauspost) [inline detect_arm64.go]: This should not be exported.

File: cpuid.go [hunk 4]
```diff
@@ -178,6 +235,17 @@ func init() {
 	Detect()
 }
 
+// isArm64Arch returns true if the architecture is ARM64.
+//
+// The arch value should be normalized before being passed to this function.
+func isArm64Arch() bool {
```
Comment (klauspost) [inline cpuid.go]: It should be possible to make this static, ie. set by a build constraint.

**Comment by lubinsz** [comment]
Hi @klauspost 
a) I would like to expose the boolean value returned by isArm64Arch. Maybe by adding ARM64 as a "Vendor".
-------------
The vendor for arm64 should be Cavium, Qualcomm, Marvell..
I will update the vendor for Arm64 in another patch.

Also can I add "cpuFamily" in the struct of "CPUInfo" ?

b) Is is possible to populate more of "CPU" variable?
----
In Arm64 platform, all "CPU" variables can be get by 'midr register' & 'hwcap'.
Please see link(http://infocenter.arm.com/help/topic/com.arm.doc.100235_0003_00_en/arm_cortex_m33_dgug_100235_0003_00_en.pdf) as reference.
Such as: 'CPU implementer' in 'midr register' means cpu vendor.
I have updated detect_arm64.go, please check it.

c) Please add tests for valid/invalid different inputs to "decodeCpuInfo". I really don't want this to blow up.
-----
I have removed it

d) Please add some documentation/examples on how to use it.
-----
I will update it in another patch :)

**Comment by lubinsz** [comment]
Hi @klauspost ,
Can you take a look at it？
Thanks

**Comment by lubinsz** [comment]
Hi @klauspost ,
Can you take a look at it？
Thanks

**Comment by fwessels** [comment]
Superseded by PR #44

**Pull request closed without merging**